src/System.Sorting/src/TimSort.cs

```diff
@@ -0,0 +1,680 @@
+        //MinMerge number above. we may need to readjust these nos, in c# based on code performance.
+        internal const int StackLength = 40;
+
+        //internal T[] mergeBufferArray=new T[256];
+
+        //The minimum run size (minrun) depends on the size of the array.
+        //The final algorithm takes the six most significant bits of the size of the array, adds one if any of the remaining bits are set, and uses that result as the minrun.This algorithm works for all arrays, including those smaller than 32
+        internal static int FindMinimumRun(int arrayLength)
+        {
+            int remainder = 0;  /* becomes 1 if the least significant bits contain at least one off bit */
+            while (arrayLength >= MinMerge)
+            {
+                remainder |= arrayLength & 1;
+                arrayLength >>= 1;
+            }
+            return arrayLength + remainder;
+        }
+
+    }
+    //public class TimSort<T> where T : IComparable<T>
+    public class TimSort<T> where T : IComparable,new()
+    {
+        //counter to increase the arrayRunBase & arrayRunLength
+        private int stackSize;
+
+        /// <summary>
+		/// A stack of pending runs yet to be merged. 
+        /// Run i starts at
+		/// address base[i] and extends for len[i] elements.  It's always
+		/// true (so long as the indices are in bounds) that:
+		/// <c>runBase[i] + runLen[i] == runBase[i + 1]</c>
+        private int[] arrayRunBase;
+        private int[] arrayRunLength;
+
+        /// <summary>
+        /// Temp storage for merges.
+        /// </summary>
+        private T[] mergeBufferArray;// = new T[256];
+        
+        /// <summary>The array being sorted.</summary>
+		private readonly T[] array;
+        private int m_minGallop;
+
+        /*** The comparer for this sort.*/
+        private IComparer<T> vComparable;
+
+        //private TimSort(T[] keys, IComparer<T> comparer)
+        public TimSort(T[] keys, IComparer<T> comparer)
+        {
+            array = keys;
+            arrayRunBase = new int[TimSortUtilities.StackLength];
+            arrayRunLength = new int[TimSortUtilities.StackLength];
+            vComparable = comparer;
+            m_minGallop = TimSortUtilities.MinGallop;
+            mergeBufferArray = new T[array.Length];
+        }
+
+        //static volatile TimSort<T> defaultArraySortHelper;
+        //public static TimSort<T> Default(T[] keys, IComparer<T> comparer)
+        //{
+        //    //get{
+        //        //TimSort<T> sorter = defaultArraySortHelper;
+        //        if (defaultArraySortHelper == null)
+        //        defaultArraySortHelper = new TimSort<T>(keys, comparer);
+
+        //        return defaultArraySortHelper;
+        //    //}
+        //}
+
+        public void Sort(T[] keys, int index, int arrayLength, IComparer<T> comparer)
+        {
+            Contract.Assert(keys != null, "Check the arguments in the caller!");
```

Comment:
We shouldn't put `Contract.Assert`s on public methods, instead we should actually throw ArgumentNullException, or ArgumentOutOfRangeException explicitly.

Internally, if you want to use asserts (which is great) please use Debug.Assert instead.